bep_webrtc.rst

```diff
@@ -0,0 +1,241 @@
+by Feross Aboukhadijeh and hundreds of open source contributors.
+
+overview
+--------
+The websocket tracker uses JSON payloads reflecting the HTTP request parameters
+and an additional action property used to switch between "announce" and other
+actions (ex. scrape). If the announce URL of the torrent contains the ws or wss
+protocol, a websocket connection is established between the client and the
+tracker.
+
+WebRTC offers and answers can be provided as an extension of the announce
+message, the tracker will be responsible to forward them between the peers to
+act as a signaling service.
+
+After the connection is open, the client can begin announcing itself by using
+the JSON messages below.
+
+signaling related message format
+---------------------------------
+announce request::
+
+  {
+    "action":     "announce",
+    "info_hash":  "",
+    "peer_id":    ""
+    "numwant":    0,
+    "uploaded":   0,
+    "downloaded": 0,
+    "left":       0,
+    "event":      "",
+    "offers":     [
+      {
+        "offer_id": "",
+        "offer": ""
+      },
+      ...
+    ]
+  }
+
+A client can provide SDP offers that the tracker will forward to peers to
+establish a connection.
+"offers" is a JSON array containing a list of WebRTC SDP offers and a generated
+"offer_id" used to match peers and offers.
+
+
+announce response::
+
+  {
+    "action":     "announce",
+    "complete":   0,
+    "incomplete": 0,
+    "interval":   0,
+    "info_hash":  ""
+  }
+
+
+offer or answer message::
+
+  {
+    "action":     "announce",
+    "info_hash":  "",
+    "offer_id":   "",
+    "peer_id":    "",
+    "sdp":      ""
+  }
+
+The tracker forwards the offer or the answer to clients in an announce message
+with an "sdp" property.
+
+
+answer message::
+
+  {
+    "action":     "announce",
+    "info_hash":  "",
+    "offer_id":   "",
+    "peer_id":    "",
+    "to_peer_id":    "",
+    "answer":     ""
+  }
+
+A client can answer to an offer by sending the data in an announce message with
+an "answer" property. The "to_peer_id" property tells the tracker to which
+peer it has to forward the message.
+
+
+other message format
+--------------------
+
+scrape request::
+
+  {
+    "action":     "scrape",
+    "info_hash":  ""
+  }
+
+
+scrape response::
+
+  {
+    "announce": "",
+    "info_hash": "",
+    "complete": 0,
+    "incomplete": 0,
+    "downloaded": 0
+  }
+
+
+multi-scrape request::
+
+  {
+    "action":     "scrape",
+    "info_hash":  ["ih1", "ih2", ...]
+  }
+
+
+multi-scrape response::
+
+  {
+    "ih1": {
+      "announce": "",
+      "info_hash": "ih1",
+      "complete": 0,
+      "incomplete": 0,
+      "downloaded": 0
+    },
+    "ih2":
+    {
+      "announce":   "",
+      "info_hash":  "ih2",
+      "complete":   0,
+      "incomplete": 0,
+      "downloaded": 0
+    }
+  }
+
+
+If the tracker encounters an error, it might send an error message.
+
+error response::
+
+  {
+    "error": ""
```

Comment:
Looking at the [UDP tracker BEP](http://www.bittorrent.org/beps/bep_0015.html), I've found it uses a type of `action` called `error` whenever an error is return. And then a `message` attribute for the error message.

It could make sense adding it (at least the `action` attribute), just to be consistent in the response.